src/rpm_ostree/cli_upgrade.rs

```diff
@@ -0,0 +1,23 @@
+//! Interface to `rpm-ostree upgrade --lock-finalization`.
+
+use super::Release;
+use failure::{bail, format_err, Fallible, ResultExt};
+
+/// Upgrade and leave the new deployment locked.
+pub fn locked_upgrade(release: Release) -> Fallible<Release> {
+    let cmd = std::process::Command::new("rpm-ostree")
+        .arg("upgrade")
+        .arg("--lock-finalization")
+        .arg(release.checksum.to_string())
+        .output()
+        .with_context(|e| format_err!("failed to run rpm-ostree: {}", e))?;
```

Comment:
@jlebon my understanding is that this is a blocking command (i.e. its exit code is reliable to signal success). If not, I may need to tweak the state machine a bit.